GH-8220: Approvals e2e tests ( regular & accessibility )

plabadie opened the pull request:
closes GH-8220

WeizhengSap committed:
feature/GH-8220: add order approval e2e test

WeizhengSap committed:
feature/GH-8220: mock order approval requests

plabadie committed:
Merge branch 'develop' into feature/GH-8220-b

plabadie committed:
GH-8220: Add Accessibility e2e first draft.

WeizhengSap committed:
feature/GH-8220: add more order approval accessibility tests

plabadie committed:
GH-8220: Fix approval accessibility e2e

plabadie committed:
GH-8220: Add tabbing order use cases.

plabadie committed:
GH-8220: Add coverage for approval list row data.

plabadie committed:
GH-8220: Work in progress.

WeizhengSap committed:
feature/GH-8220: add e2e test for order approval and rejection

WeizhengSap committed:
feature/GH-8220: add some cont vars for order apporval e2e test

WeizhengSap committed:
feature/GH-8220: update this branch

WeizhengSap committed:
feature/GH-8220: merge order approval accessibility tests into b2b checkout tests

plabadie committed:
GH-8220: Add waits to approval/rejection tests.

plabadie committed:
GH-8220: Small e2e test cleanup.

plabadie committed:
GH-8220: E2e test cleanup.

plabadie committed:
GH-8220: Complete the revert.

hackergil commented on projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts:
```diff
@@ -0,0 +1,128 @@
+export function loginB2bUser() {
+  const user: AccountData = {
+    user: '',
+    registrationData: {
```
This is duplicated below (except for one line), and also there are similar snippets in other helper files. Can we get rid of the duplications as much as possible?

hackergil commented on projects/storefrontapp-e2e-cypress/cypress/integration/b2b/accessibility/tabbing-order.e2e-spec.ts:
```diff
@@ -131,4 +141,39 @@ describe('Tabbing order for B2B checkout', () => {
+  describe('Order Approval', () => {
+    before(() => {
+      cy.window().then((win) => win.sessionStorage.clear());
+      Cypress.env('BASE_SITE', POWERTOOLS_BASESITE);
```
This is already a B2B test. Can we find a way to not set the baseSite every time?

WeizhengSap replied:
done, `Cypress.env('BASE_SITE', POWERTOOLS_BASESITE)` was removed

hackergil commented on projects/storefrontapp-e2e-cypress/cypress/integration/b2b/accessibility/tabbing-order.e2e-spec.ts:
```diff
@@ -131,4 +141,39 @@ describe('Tabbing order for B2B checkout', () => {
+
+    context('Approval List', () => {
+      it('should allow to navigate with tab key', () => {
+        loginB2bApprover();
```
Duplication on calling loginB2BApprover can be simplified placing it on a beforeEach() hook. Or even better, just once on before() hook.

WeizhengSap replied:
done, moved to `beforeEach()`.

hackergil commented on projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts:
```diff
@@ -0,0 +1,234 @@
+  });
+
+  it('Should display approval detail page', () => {
+    orderApproval.loginB2bApprover();
```
Same as above. login can be done on beforeEach() or before() hook

WeizhengSap replied:
done, move to `beforeEach()`

hackergil commented on projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts:
```diff
@@ -0,0 +1,234 @@
+  // assert products
+  cy.get('.cx-item-list-row .cx-link').should(
+    'contain',
+    sampleData.approvalOrderDetail.order.entries[0].product.name
```
Can we shorten this to
firstOrderEntry = sampleData.approvalOrderDetail.order.entries[0]
and then assert on firstOrderEntry.* ?
(law of Demeter)

hackergil commented on projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts:
```diff
@@ -0,0 +1,128 @@
+  visitHomePage();
+}
+
+export function getPendingOrderDetails() {
```
can we make more explicit in the name of these functions that we are stubbing the response of an xhr call?

plabadie replied:
Done. We update the names of all functions which use a stubbed response.

hackergil commented on projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts:
```diff
@@ -0,0 +1,234 @@
+    .eq(3)
+    .should(
+      'contain',
+      sampleData.approvalOrderDetail.order.deliveryAddress.formattedAddress
```
Same comment as above

hackergil reviewed (changes requested):
Some questions:
* I had to update my cypress file to start as b2b with the right basesite. What's the plan to handle this? Do we know?
* What happened with the idea of incorporating tabbing to the test?
* I noticed on the order approval list page there are some dropdowns that are not being tested. Why?

plabadie committed:
GH-8220: Apply code review improvements.

WeizhengSap committed:
feature/GH-8220: address review comments

WeizhengSap committed:
feature/GH-8220: update this branch

WeizhengSap committed:
feature/GH-8220: fix merge conficts

plabadie committed:
GH-8220: Move approvals accessibility e2e over to the main approvals e2e

plabadie committed:
GH-8220: Worrk in progress

WeizhengSap committed:
feature/GH-8220: test sort dropdown in order approval list page

WeizhengSap committed:
feature/GH-8220: change the order approval code in sample data

plabadie commented:
> Some questions:
> 
> * I had to update my cypress file to start as b2b with the right basesite. What's the plan to handle this? Do we know?

The yarn scripts provide the base site as an argument when they start the b2b tests.
Locally you can use some of the existing commands in the package.json or add "--env BASE_SITE=powertools-spa" to the commands you use to launch cypress. 

> * What happened with the idea of incorporating tabbing to the test?

We had already pushed this ticket for review when we had the team discussion, but we now incorporated the accessibility tests into the "regular" approvals e2e.

> * I noticed on the order approval list page there are some dropdowns that are not being tested. Why?

It is a shared component that drives this functionality, so I guess we left it out initially.  However, we now added coverage for the sorting dropdown.

hackergil commented on projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts:
```diff
@@ -0,0 +1,275 @@
+  });
+
+  describe('Order Approval - Accessibility.', () => {
+    beforeEach(() => {
```
Logging in consumes some seconds. The pattern that we are following in other tests is:
- Log in once
- Store the token in local storage
- restore local storage before each test, to get the token
That way, you don't have to login again. You'll reduce the execution time for at least 30 seconds by logging in once or twice as an approver instead of 8 times (before each test)

```suggestion
    before(() => {
      orderApproval.loginB2bApprover();
      cy.saveLocalStorage();
    });

    beforeEach(() => {
      cy.restoreLocalStorage();
    });
```

plabadie replied:
Done.

hackergil commented on projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts:
```diff
@@ -0,0 +1,133 @@
+
+export function loginB2bUser() {
+  cy.requireLoggedIn(sampleData.b2bUserAccount);
+  visitHomePage();
```
I think you don't need to visit the homepage here. This takes about 3 seconds.

plabadie replied:
Done.

hackergil commented on projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts:
```diff
@@ -0,0 +1,133 @@
+
+export function loginB2bApprover() {
+  cy.requireLoggedIn(sampleData.b2bApproverAccount);
+  visitHomePage();
```
I think you don't need to visit the homepage here. This takes about 3 seconds.

plabadie replied:
Done.

hackergil commented on projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts:
```diff
@@ -0,0 +1,275 @@
+      orderApproval.visitOrderApprovalListPage();
+      cy.wait('@order_approval_list').its('status').should('eq', 200);
+
+      cy.get('cx-order-approval-list a.cx-order-approval-value')
```
Maybe there is a way to abstract these selectors into either a function or a block so that we don't repeat them?

plabadie replied:
Done.

hackergil commented on projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts:
```diff
@@ -0,0 +1,275 @@
+    cy.window().then((win) => win.sessionStorage.clear());
+  });
+
+  describe('Order Approval - Accessibility.', () => {
```
My idea of including the accessibility tests within the main test was to have one test only that validated both the user tabbing and also the functionality (not to place the two tests/journeys in a single file). I checked how the tabbing helpers work and at the moment it is not possible. I apologize for not explaining my intention clearly.

Because of this, I think it's better if the accessibility tests are placed under b2b/accessibility on their own file.

plabadie committed:
GH-8220: Code review improvements

plabadie committed:
GH-8220: Incorporate accessibilty tests in the regular tests.

plabadie committed:
Merge branch 'develop' into feature/GH-8220-b

plabadie committed:
Merge branch 'develop' into feature/GH-8220-b

plabadie committed:
Merge branch 'develop' into feature/GH-8220-b

WeizhengSap committed:
Merge branch 'develop' into feature/GH-8220-b

WeizhengSap committed:
Merge branch 'develop' into feature/GH-8220-b

WeizhengSap committed:
Merge branch 'develop' into feature/GH-8220-b

WeizhengSap committed:
Merge branch 'develop' into feature/GH-8220-b

WeizhengSap committed:
Merge branch 'develop' into feature/GH-8220-b

WeizhengSap committed:
Merge branch 'develop' into feature/GH-8220-b
